pkg/component/types.go

```diff
@@ -71,3 +80,146 @@ const (
 	// StateTypeUnknown means that odo cannot tell its state
 	StateTypeUnknown State = "Unknown"
 )
+
+func newComponentWithType(componentName, componentType string) Component {
+	cmp := NewComponent(componentName)
+	cmp.Spec.Type = componentType
+	return cmp
+}
+
+// NewComponent provides a constructor to component struct with some metadata prefilled
+func NewComponent(componentName string) Component {
+	return Component{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       ComponentKind,
+			APIVersion: machineoutput.APIVersion,
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name: componentName,
+		},
+		Status: ComponentStatus{},
+	}
+}
+
+// newComponentList returns list of devfile and s2i components in machine readable format
+func newComponentList(s2iComps []Component) ComponentList {
+	if len(s2iComps) == 0 {
+		s2iComps = []Component{}
+	}
+
+	return ComponentList{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       machineoutput.ListKind,
+			APIVersion: machineoutput.APIVersion,
+		},
+		ListMeta: metav1.ListMeta{},
+		Items:    s2iComps,
+	}
+}
+
+// NewCombinedComponentList returns list of devfile, s2i components and other components(not managed by odo) in machine readable format
+func NewCombinedComponentList(s2iComps []Component, devfileComps []Component, otherComps []Component) CombinedComponentList {
+	if len(s2iComps) == 0 {
+		s2iComps = []Component{}
+	}
+	if len(devfileComps) == 0 {
+		devfileComps = []Component{}
+	}
+	if len(otherComps) == 0 {
+		otherComps = []Component{}
+	}
+
+	return CombinedComponentList{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       machineoutput.ListKind,
+			APIVersion: machineoutput.APIVersion,
+		},
+		ListMeta:          metav1.ListMeta{},
+		S2IComponents:     s2iComps,
+		DevfileComponents: devfileComps,
+		OtherComponents:   otherComps,
+	}
+}
+
+func (o CombinedComponentList) HumanReadableOutputInPath(wr io.Writer) {
```

Comment:
I feel this method should be in the CLI layer

Reply:
You are right, I'll move these methods to the CLI layer